README.md
# pwseqdist

[![Build Status](https://travis-ci.com/kmayerb/pwseqdist.svg?branch=master)](https://travis-ci.com/kmayerb/pwseqdist)
[![PyPI version](https://badge.fury.io/py/pwseqdist.svg)](https://badge.fury.io/py/pwseqdist)
[![Coverage Status](https://coveralls.io/repos/github/agartland/pwseqdist/badge.svg?branch=master)](https://coveralls.io/github/agartland/pwseqdist?branch=master)

A small package that efficiently computes distances between protein sequences.
Can accommodate similarity matrices, sequences of different lengths and custom
metrics.

## Install

```
pip install pwseqdist
```

## Example

```python
import pwseqdist as pw
import multiprocessing
from scipy.spatial.distance import squareform

peptides = ['CACADLGAYPDKLIF','CACDALLAYTDKLIF',
            'CACDAVGDTLDKLIF','CACDDVTEVEGDKLIF',
            'CACDFISPSNWGIQSGRNTDKLIF','CACDPVLGDTRLTDKLIF']

dvec = pw.apply_pairwise_sq(seqs = peptides, 
			    metric = pw.metrics.nw_hamming_metric, 
			    ncpus  = multiprocessing.cpu_count() )

dmat = squareform(dvec).astype(int)
dmat
```

```
array([[ 0,  4,  6,  7, 15,  8],
       [ 4,  0,  5,  7, 14,  7],
       [ 6,  5,  0,  6, 14,  4],
       [ 7,  7,  6,  0, 14,  8],
       [15, 14, 14, 14,  0, 11],
       [ 8,  7,  4,  8, 11,  0]])
```


HISTORY.md
# Release History 

## 0.1.0 2020-05-18 # Development Release

pwseqdist/tests/test_examples.py
import pytest 

def test_README_example1():
	import numpy as np

	import pwseqdist as pw
	import multiprocessing
	from scipy.spatial.distance import squareform

	peptides = ['CACADLGAYPDKLIF','CACDALLAYTDKLIF',
	            'CACDAVGDTLDKLIF','CACDDVTEVEGDKLIF',
	            'CACDFISPSNWGIQSGRNTDKLIF','CACDPVLGDTRLTDKLIF']

	dvec = pw.apply_pairwise_sq(seqs   = peptides, 
								metric = pw.metrics.nw_hamming_metric, 
								ncpus  = multiprocessing.cpu_count()   )

	dmat = squareform(dvec).astype(int)
	
	exp = np.array(
	   [[ 0,  4,  6,  7, 15, 8],
       [ 4,  0,  5,  7, 14,  7],
       [ 6,  5,  0,  6, 14,  4],
       [ 7,  7,  6,  0, 14,  8],
       [15, 14, 14, 14,  0, 11],
       [ 8,  7,  4,  8, 11,  0]])

	assert np.all(dmat == exp)

pwseqdist/tests/test_matrices.py
import pwseqdist as pwsd
import parasail
import numpy as np
import itertools
import pytest

def alphabet_oder():
	assert pwsd.matrices.parasail_aa_alphabet == 'ARNDCQEGHILKMFPSTWYVBZX'

def test_dict_from_matrix():

	d = pwsd.matrices.dict_from_matrix(parasail.blosum62)
	assert isinstance(d, dict)

def test_dict_from_matrix_against_external_blossum_reference():
	# See truth values: https://www.ncbi.nlm.nih.gov/books/NBK6831/figure/A551/
	# ftp://ftp.ncbi.nlm.nih.gov/blast/matrices
	d = pwsd.matrices.dict_from_matrix(parasail.blosum62)
	assert isinstance(d, dict)
	assert d[('X', 'V')] == -1
	assert d[('A', 'A')] == 4
	assert d[('N', 'N')] == 6
	assert d[('W', 'W')] == 11
	assert d[('D', 'W')] == -4
	assert d[('W', 'D')] == -4
	# No diagnol entries less than 4
	assert np.all([d[(i, i)] >=4 for i in  pwsd.matrices.parasail_aa_alphabet if i != "X"] )
	# No off diagnol entries greater than 4
	assert np.all([d[(i[0], i[1])] <=4 for i in \
	 itertools.product(pwsd.matrices.parasail_aa_alphabet, pwsd.matrices.parasail_aa_alphabet) if i[0] !=i[1]] )

def test_seq2veq():
	assert(isinstance(pwsd.matrices.seq2vec("CAT"), np.ndarray))

def test_seq2veq():
	assert np.all(pwsd.matrices.seq2vec("ARNDCQEGHILKMFPSTWYVBZX") == \
		np.array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,17, 18, 19, 20, 21, 22]))

def test_vec2seq():
    assert pwsd.matrices.vec2seq([ 4,  0, 16]) == 'CAT'

def test_vec2seq_out_of_range_alph24():
	with pytest.raises(ValueError) as ex:
		pwsd.matrices.vec2seq([ 4,  0, 24], alphabet = pwsd.matrices.parasail_aa_alphabet_with_unknown )
	print(ex.value)
	assert str(ex.value) == 'vex2seq only works with integers 0 to 23 corresponding with ARNDCQEGHILKMFPSTWYVBZX*'

def test_vec2seq_out_of_range_alph23():
	with pytest.raises(ValueError) as ex:
		pwsd.matrices.vec2seq([ 4,  0, 23], alphabet = pwsd.matrices.parasail_aa_alphabet )
	print(ex.value)
	assert str(ex.value) == 'vex2seq only works with integers 0 to 22 corresponding with ARNDCQEGHILKMFPSTWYVBZX'


def test_seqs2mat():
	assert np.all(pwsd.matrices.seqs2mat(["CAT","HAT"]) == np.array([[ 4,  0, 16],[ 8,  0, 16]]))
    
def test_mat2seqs():
	assert np.all(pwsd.matrices.mat2seqs(np.array([[ 4,  0, 16],[ 8,  0, 16]]) ) == ['CAT', 'HAT'])


pwseqdist/tests/test_pairwise.py
import pwseqdist as pwsd
import parasail
import numpy as np
import itertools
import pytest

mixed_seqs = ['CACADLGAYPDKLIF',
         'CACDALLAYTDKLIF',
         'CACDAVGDTLDKLIF',
         'CACDDVTEVEGDKLIF',
         'CACDFISPSNWGIQSGRNTDKLIF',
         'CACDILLGDTADKLIF',
         'CACDIVLSGGLDTRQMFF',
         'CACDLLLRQSSTDKLIF',
         'CACDNLSETTDKLIF',
         'CACDPLGTDKLIF',
         'CACDPMGGSGGLSWDTRQMFF',
         'CACDPVLGDTRLTDKLIF',
         'CACDPVQGYSGQNRAYTDKLIF',
         'CACDSILGDTLYTDKLIF',
         'CACDSLTSHTGGFGPDKLIF',
         'CACDSTGDLSSWDTRQMFF',
         'CACDSVESRNVLGDPTTDKLIF',
         'CACDSVLSRDLGDSELIF',
         'CACDTAAGGYASSWDTRQMFF',
         'CACDTAPHGGRTWDTRQMFF',
         'CACDTGGYVNWDTRQMFF',
         'CACDTGRLLGDTADTRQMFF',
         'CACDTIRGFSSWDTRQMFF',
         'CACDTIVAPALDKLIF',
         'CACDTLFLGEDTPTDKLIF',
         'CACDTLGDLSLTAQLFF',
         'CACDTLGDPPHTDKLIF',
         'CACDTLGDYTQSDKLIF',
         'CACDTLGGYPWDTRQMFF',
         'CACDTLGKTDKLIF',
         'CACDTLPLKTGGPLYTDKLIF',
         'CACDTLRLGDPLNTDKLIF',
         'CACDTVALGDTESSWDTRQMFF',
         'CACDTVGAVLGDPKGTDKLIF',
         'CACDTVGDGPDTDKLIF',
         'CACDTVGDTADKLIF',
         'CACDTVGDTHSWDTRQMFF',
         'CACDTVGGSTDKLIF',
         'CACDTVGIPPDKLIF',
         'CACDTVGYGEGDTDKLIF',
         'CACDTVISSNRRGGDKLIF',
         'CACDTVPPGDTGTDKLIF',
         'CACDTVRFTGGYENTDKLIF',
         'CACDYVLGAEDKLIF',
         'CACEGILKSEPLGIDKLIF',
         'CACEMLGHPPGDKLIF',
         'CACVSLDLSYTDKLIF',
         'CALGEIAFRSRTGGPPYTDKLIF',
         'CALGTAYFLRDPGADKLIF',
         'CAVKVPLTSSPREGPTVLHDKLIF']


def test_nw_cpu1_no_kwargs():
	r = pwsd.pairwise.apply_pairwise_sq(seqs   = mixed_seqs, 
										metric = pwsd.metrics.nw_metric,
										ncpus = 1)
	assert isinstance(r, np.ndarray)

expectation = pwsd.pairwise.apply_pairwise_sq(seqs   = mixed_seqs, 
										metric = pwsd.metrics.nw_metric,
										ncpus = 1)
testspace = [
	(mixed_seqs, pwsd.metrics.nw_metric, 1,  expectation, None),
	(mixed_seqs, pwsd.metrics.nw_metric, 2,  expectation, None),
	(mixed_seqs, pwsd.metrics.nw_metric, 4,  expectation, None),
	(mixed_seqs, pwsd.metrics.nw_metric, 8,  expectation, None),
	(mixed_seqs, pwsd.metrics.nw_metric, 32, expectation, None),
	(mixed_seqs, pwsd.metrics.hamming_distance, 1, expectation, AssertionError)]

@pytest.mark.parametrize("s,d,cpus,expected,expected_error", testspace)
def test_nw_multcpu_no_kwargs(s,d,cpus,expected,expected_error):
	try:
		r = pwsd.pairwise.apply_pairwise_sq(seqs   = s, 
											metric = d,
											ncpus = cpus)
		assert np.all(r == expected)
	except: 
		with pytest.raises(expected_error):
			r = pwsd.pairwise.apply_pairwise_sq(seqs   = s, 
												metric = d,
												ncpus = cpus)





def test_nw_hamming_cpu1_no_kwargs():
	r = pwsd.pairwise.apply_pairwise_sq(seqs   = mixed_seqs, 
										metric = pwsd.metrics.nw_metric,
										ncpus = 1)
	assert isinstance(r, np.ndarray)


expectation = pwsd.pairwise.apply_pairwise_sq(seqs   = mixed_seqs, 
											  metric = pwsd.metrics.nw_hamming_metric,
											  ncpus = 1)
testspace = [
	(mixed_seqs, pwsd.metrics.nw_hamming_metric, 1,  expectation, None),
	(mixed_seqs, pwsd.metrics.nw_hamming_metric, 2,  expectation, None),
	(mixed_seqs, pwsd.metrics.nw_hamming_metric, 4,  expectation, None),
	(mixed_seqs, pwsd.metrics.nw_hamming_metric, 8,  expectation, None),
	(mixed_seqs, pwsd.metrics.nw_hamming_metric, 32, expectation, None)]
@pytest.mark.parametrize("s,d,cpus,expected,expected_error", testspace)
def test_first_nw_hamming_metric_multcpu_no_kwargs(s,d,cpus,expected,expected_error):
	"""
	Tests that output is the same for a range of cpus
	"""
	try:
		r = pwsd.pairwise.apply_pairwise_sq(seqs   = s, 
											metric = d,
											ncpus = cpus)
		assert np.all(r == expected)
	except: 
		with pytest.raises(expected_error):
			r = pwsd.pairwise.apply_pairwise_sq(seqs   = s, 
												metric = d,
												ncpus = cpus)



pwseqdist/metrics.py
import parasail
import numpy as np
import operator

__all__ = ['compute_many',
           'compute_many_rect',
           'nw_metric',
           'nw_hamming_metric',
           'np_subst_metric',
           'str_subst_metric']

def compute_many(indices, metric, seqs, dtype, **kwargs):
    return  np.array([metric(seqs[i], seqs[j], **kwargs) for i,j in indices], dtype=dtype)

def compute_many_rect(indices, metric, seqs1, seqs2, dtype, **kwargs):
    return np.array([metric(seqs1[i], seqs2[j], **kwargs) for i,j in indices], dtype=dtype)

def _str_sim(s1, s2, subst, na_penalty):
    a = np.array([i for i in map(lambda a, b: subst.get((a, b), subst.get((b, a), na_penalty)), s1, s2)])
    return np.sum(a)

def str_subst_metric(seq1, seq2, subst_dict, as_similarity=False, na_penalty=None):
    """Distance between seq1 and seq2 using a substitution similarity matrix.
    Requires that seq1 and seq2 have the same length (see nw_metric for alternative)
    
    Distance is a function of these similarities:
    
        D = sim11 + sim22 - 2 * sim12

    Parameters
    ----------
    seq1: string
        string containing amino acid letters
    seq2: string
        string containing amino acid letters
    subst_dict : dict
        A substitution dictionary with keys (AA1, AA2).
        Can be created from any parasail matrix by calling
        pwseqdist.matrices.dict_from_matrix,
        e.g.  subst_dict = dict_from_matrix(parasail.blosum62)
    as_similarity : bool
        Returns values as summed raw, similarities as opposed to distances.
    na_penalty : int
        Optionally provide a penalty for any AA symbols not in the parasail
        alphabet: ARNDCQEGHILKMFPSTWYVBZX. This value is also coded in 
        matrices that are translated from parasail.

    Returns
    -------
    D : float
        distance between strings"""
    msg = "len of seq1 (%d) and seq2 (%d) are different" % (len(seq1), len(seq2))
    assert len(seq1) == len(seq2), msg
    
    if na_penalty is None:
        na_penalty = subst_dict['na']

    """Site-wise similarity between seq1 and seq2 using the substitution matrix subst"""
    sim12 = _str_sim(seq1, seq2, subst_dict, na_penalty)

    if as_similarity:
        return sim12
    else:
        L = len(seq1)
        sim11 = _str_sim(seq1, seq1, subst_dict, na_penalty)
        sim22 = _str_sim(seq2, seq2, subst_dict, na_penalty)
        D = sim11 + sim22 - 2 * sim12
        return D

def hamming_distance(s1, s2):
    """Hamming distance between str1 and str2.
    Requires that str1 and sr2 are equal length

    Parameters
    ----------
    s1: string
        string containing amino acid letters
    s2: string
        string containing amino acid letters

    Returns
    -------
    D : float
        distance between strings (Hamming Distance: number of mismatched positions)"""
    assert len(s1) == len(s2), "Inputs must have the same length."
    return np.sum([i for i in map(operator.__ne__, s1, s2)])

def nw_metric(s1, s2, matrix='blosum62', open=3, extend=3):
    """Function applying Parasail's Needleman-Wuncsh Algorithm to compute
    a distance between any two sequences.

    Parameters
    ----------
    s1: string
        string containing amino acid letters
    s2: string
        string containing amino acid letters
    matrix : str
        Attribute of parasail that names a substitution matrix

    Returns
    -------
    D : float
        distance via reciprocal alignment scores.

    Notes
    -----

    .. code-block:: python

      xx = parasail.nw_stats(s1, s1, open=open, extend=extend, matrix=matrix).score
      yy = parasail.nw_stats(s2, s2, open=open, extend=extend, matrix=matrix).score
      xy = parasail.nw_stats(s1, s2, open=open, extend=extend, matrix=matrix).score
      D = xx + yy - 2 * xy
      return D


    May or may not produce a true metric. Details in:
    E. Halpering, J. Buhler, R. Karp, R. Krauthgamer, and B. Westover.
    Detecting protein sequence conservation via metric embeddings.
    Bioinformatics, 19 (sup 1) 2003
    """
    p_matrix = getattr(parasail, matrix)
    xx = parasail.nw_stats(s1, s1, open=open, extend=extend, matrix=p_matrix).score
    yy = parasail.nw_stats(s2, s2, open=open, extend=extend, matrix=p_matrix).score
    xy = parasail.nw_stats(s1, s2, open=open, extend=extend, matrix=p_matrix).score
    D = xx + yy - 2 * xy
    return D

def nw_hamming_metric(s1, s2, matrix='blosum62', open=3, extend=3):
    """Function applying Parasail's Needleman-Wuncsh Algorithm to align and
    compute a Hamming Distance between any two sequences: number of
    mismatched positions. Gaps count as a mismatch. Penalties and matrix
    are used for alignment purposes, not in the distance calculation.

    Parameters
    ----------
    s1: string
        string containing amino acid letters
    s2: string
        string containing amino acid letters
    matrix : str
        Attribute of parasail that names a substitution matrix

    Returns
    -------
    D : float
        distance between strings (Hamming Distance: number of mismatched positions)

    Notes
    -----
    .. code-block:: python

        xy = parasail.nw_stats(s1, s2, open=open, extend=extend, matrix=matrix)
        xy_t = parasail.nw_trace(s1, s2, open=open, extend=extend, matrix=matrix)
        hamming_distance = len(xy_t.traceback.comp)-xy.matches
        return hamming_distance"""
    p_matrix = getattr(parasail, matrix)
    xy = parasail.nw_stats(s1, s2, open=open, extend=extend, matrix=p_matrix)
    xy_t = parasail.nw_trace(s1, s2, open=open, extend=extend, matrix=p_matrix)
    D = len(xy_t.traceback.comp)-xy.matches
    return D

def np_subst_metric(seq1, seq2, subst_mat, as_similarity=False):
    """Numpy version of str_subst_metric, however it is not yet compatible with
    apply_pairwise_sq because of the need to convert strings to vectors of integers.
    It may not be faster than the string version, so there is no plan to further develop."""
    """Computes sequence similarity or distance based on the substitution matrix."""
    msg = "Sequences must be the same length (%d != %d)." % (seq1.shape[0], seq2.shape[0])
    assert seq1.shape[0] == seq2.shape[0], msg

    """Similarity between seq1 and seq2 using the substitution matrix subst"""
    sim12 = np.sum(subst_mat[seq1, seq2])
    if as_similarity:
        return sim12
    else:
        L = seq1.shape[0]
        sim11 = np.sum(subst_mat[seq1, seq1])
        sim22 = np.sum(subst_mat[seq2, seq2])
        D = sim11 + sim22 - 2 * sim12
        return D

pwseqdist/pairwise.py
import multiprocessing
import parmap
import itertools
import numpy as np
import scipy

from .metrics import compute_many, compute_many_rect

__all__ = ['apply_pairwise_sq',
           'apply_pairwise_rect']

"""TODO:
Currently I pass all the sequences and some set of indices to compute_many. 
Why wouldn't I just send the some of the sequences?
The point was to avoid sending all the pairs of sequences and just send
pairs of indices to the workers. So we'd have to be smart about reducing the
total number of sequences that are needed and then sending just those and
translated indices.


These functions are currently not compatible with the numpy subst_metric
because the seqs input woudl be provided as a numpy matric of integers.
I'm not even sure the numpy metric will be faster so not adding now.

Add a numb-compiled version of pairwise_sq and pairwise_rect
  - Convert metrics to code that can be compile with numba
    This would allow the "outer" loop performing pairwise
    distances to also be compiled using numba. The advantage
    is that numba-compiled code can use multithreading to run
    on multiple CPUs, and making use of shared memory.

    I think this could allow for very fast distance calculations"""

def _mati2veci(i, j, n):
    veci = scipy.special.comb(n, 2) - scipy.special.comb(n - i, 2) + (j - i - 1)
    return int(veci)

def apply_pairwise_sq(seqs, metric, ncpus=1, **kwargs):
    """Calculate distance between all pairs of seqs using metric
    and kwargs provided to metric. Will use multiprocessing Pool
    if ncpus > 1.

    For efficiency, will only compute metric on unique values in
    seqs. All values are returned, including redundancies.

    Though written to be used for distance calculations,
    it is general enough that it could be used to run
    any arbitrary function on pairs of elements in seqs.

    Parameters
    ----------
    seqs : list
        List of sequences provided to metric in pairs.
    metric : function
        A distance function of the form
        func(seq1, seq2, **kwargs)
    ncpus : int
        Size of the worker pool to be used by multiprocessing
    **kwargs : keyword arguments
        Additional keyword arguments are supplied to the metric.

    Returns
    -------
    dvec : np.ndarray, length n*(n - 1) / 2
        Vector form of the pairwise distance matrix.
        Use scipy.distance.squareform to convert to a square matrix"""
    """Set to false to turn on computation of redundant distances"""
    uniqify=True

    useqs = list(set(seqs))
    if len(useqs) == len(seqs) or ~uniqify:
        useqs = seqs
        translate = False
    else:
        translate = True

    """itertools.combinations creates the i,j pairs in the same order
    as scipy.distance.pdist/squareform"""
    pw_indices = list(itertools.combinations(range(len(useqs)), 2))

    chunk_func = lambda l, n: [l[i:i + n] for i in range(0, len(l), n)]
    chunksz = len(pw_indices) // ncpus
    chunked_indices = chunk_func(pw_indices, chunksz)
    dtype = type(metric(useqs[0], useqs[0], **kwargs))

    if ncpus > 1:
        with multiprocessing.Pool(ncpus) as pool:
            try:
                dists = parmap.map(compute_many,
                                   chunked_indices,
                                   metric,
                                   useqs,
                                   dtype,
                                   **kwargs,
                                   pm_parallel=True,
                                   pm_pool=pool)
            except ValueError as err:
                print('pwseqdist.apply_pairwise_sq: error with metric %s and multiprocessing, trying on single core' % metric)
                dists = parmap.map(compute_many,
                                   chunked_indices,
                                   metric,
                                   useqs,
                                   dtype,
                                   **kwargs,
                                   pm_parallel=False)
                print('pwseqdist.apply_pairwise_sq: metric %s could not be spread to multiple processes, ran on single core' % metric)
    else:
        dists = parmap.map(compute_many,
                               chunked_indices,
                               metric,
                               useqs,
                               dtype,
                               **kwargs,
                               pm_parallel=False)
    
    """Create translation dict from vector_i to mat_ij coordinates for the
    distance matrix of unique seqs (unneccessary, but may be useful later)"""
    # uveci2umati = {veci:(i, j) for veci, (i, j) in enumerate(itertools.combinations(range(len(useqs)), 2))}

    """Get the vector form of the useqs"""
    """n = len(useqs)
    uvec = np.zeros(scipy.special.comb(n, 2))
    for ichunk, dchunk in zip(chunked_indices, dists):
        for (i,j), d in zip(ichunk, dchunk):
            uvec[_mati2veci(i, j, n)] = d"""
    uvec = np.concatenate(dists) # this may be more memory intensive, but should be fine
    
    if translate:
        """Then translate the vector form of the useqs to the vector
        form of the seqs"""
        vout = np.zeros(int(scipy.special.comb(len(seqs), 2)))
        for veci, (i,j), in enumerate(itertools.combinations(range(len(seqs)), 2)):
            ui = useqs.index(seqs[i])
            uj = useqs.index(seqs[j])
            if ui == uj:
                vout[veci] = 0
            else:
                if uj < ui:
                    uj, ui = ui, uj
                vout[veci] = uvec[_mati2veci(ui, uj, len(useqs))]
    else:
        vout = uvec
    return vout

def apply_pairwise_rect(seqs1, seqs2, metric, ncpus=1, **kwargs):
    """Calculate distance between pairs of sequences in seqs1
    with sequences in seqs2 using metric and kwargs provided to
    metric. Will use multiprocessing Pool if ncpus > 1.

    For efficiency, will only compute metric on unique values in
    seqs1/seqs2. All values are returned, including redundancies.

    Though written to be used for distance calculations,
    it is general enough that it could be used to run
    any arbitrary function on pairs of elements in seqs.

    Parameters
    ----------
    seqs1, seqs2 : lists
        Lists of sequences.
    metric : function
        A distance function of the form
        func(seq1, seq2, **kwargs)
    ncpus : int
        Size of the worker pool to be used by multiprocessing
    **kwargs : keyword arguments
        Additional keyword arguments are supplied to the metric.

    Returns
    -------
    dvec : np.ndarray, length len(seqs1) * len(seqs2)
        Vector form of the pairwise distance rectangle.
    indices : np.ndarray, shape [len(seqs1) * len(seqs2), 2]
        Contains i,j indices on each row where i (j) is an index
        into seqs1 (seqs2) and can be used to recreate a distance rectangle"""
    uniqify = True
    def _recti2veci(i, j, n2):
        """Translate from rectangle coordinates to vector coordinates"""
        return int(i * len(n2) + j)

    useqs1 = list(set(seqs1))
    if len(useqs1) == len(seqs1) or ~uniqify:
        useqs1 = seqs1
        translate1 = False
    else:
        translate1 = True
        useqs1 = list(seqs1)

    useqs2 = list(set(seqs2))
    if len(useqs2) == len(seqs2) or ~uniqify:
        useqs2 = seqs2
        translate2 = False
    else:
        translate2 = True
        useqs2 = list(seqs2)

    pw_indices = list(itertools.product(range(len(useqs1)), range(len(useqs2))))

    chunk_func = lambda l, n: [l[i:i + n] for i in range(0, len(l), n)]
    chunksz = len(pw_indices)//ncpus
    chunked_indices = chunk_func(pw_indices, chunksz)
    dtype = type(metric(useqs1[0], useqs2[0], **kwargs))
    if ncpus > 1:
        with multiprocessing.Pool(ncpus) as pool:
            try:
                dists = parmap.map(compute_many_rect,
                                   chunked_indices,
                                   metric,
                                   useqs1,
                                   useqs2,
                                   dtype,
                                   **kwargs,
                                   pm_parallel=True,
                                   pm_pool=pool)
            except ValueError as err:
                print('pwseqdist.apply_pairwise_rect: error with metric %s and multiprocessing, trying on single core' % metric)
                dists = parmap.map(compute_many_rect,
                                   chunked_indices,
                                   metric,
                                   useqs1,
                                   useqs2,
                                   dtype,
                                   **kwargs,
                                   pm_parallel=False)
                print('pwseqdist.apply_pairwise_rect: metric %s could not be spread to multiple processes, ran on single core' % metric)
    else:
        dists = parmap.map(compute_many_rect,
                               chunked_indices,
                               metric,
                               useqs1,
                               useqs2,
                               dtype,
                               **kwargs,
                               pm_parallel=False)
    
    urect = np.concatenate(dists).reshape((len(useqs1), len(useqs2)))
    if translate1:
        redup_ind = [useqs1.index(s) for s in seqs1]
        urect = urect[redup_ind, :]
    if translate2:
        redup_ind = [useqs2.index(s) for s in seqs2]
        urect = urect[:, redup_ind]
    return urect

pwseqdist/numba_tools.py
import numpy as np
import scipy.special
import itertools

from .matrices import dict_from_matrix

__all__ = ['nb_pairwise_sq',
            'nb_hamming_distance',
            'nb_subst_metric',
            'nb_dict_from_matrix']

try:
    import numba
    # print('pwseqdist: Successfully imported numba version %s' % (numba.__version__))
    NB_SUCCESS = True
except ImportError:
    NB_SUCCESS = False
    print('pwseqdist: Could not import numba')

def nb_dict_from_matrix(subst):
    """Converts a parasail substitution matrix into a numba.typed.Dict
    that can be used with he numba metrics"""
    subst_dict = numba.typed.Dict.empty(key_type=numba.types.unicode_type,
                               value_type=numba.types.int32)
    for k,v in dict_from_matrix(subst).items():
        key = '|'.join(k)
        subst_dict[key] = v
    return subst_dict

@numba.jit(nopython=True, parallel=False)
def distance_vec(dvec, indices, seqs, nb_metric, *args):
    for veci in numba.prange(len(indices)):
        si = seqs[indices[veci, 0]]
        sj = seqs[indices[veci, 1]]
        d = nb_metric(si, sj, *args)
        dvec[veci] = d

def nb_pairwise_sq(seqs, nb_metric, *args):
    """Calculate distance between all pairs of seqs using metric
    and kwargs provided to nb_metric. Will use multiprocessing Pool
    if ncpus > 1.

    nb_metric must be a numba-compiled function

    Parameters
    ----------
    seqs : list
        List of sequences provided to metric in pairs.
    metric : numba-compiled function
        A distance function of the form
        func(seq1, seq2, **kwargs)
    **kwargs : keyword arguments
        Additional keyword arguments are supplied to the metric.

    Returns
    -------
    dvec : np.ndarray, length n*(n - 1) / 2
        Vector form of the pairwise distance matrix.
        Use scipy.distance.squareform to convert to a square matrix"""
    nb_seqs = numba.typed.List()
    for s in seqs:
        nb_seqs.append(s)

    dvec = np.zeros(int(scipy.special.comb(len(seqs), 2)))
    indices = np.zeros((int(scipy.special.comb(len(seqs), 2)), 2), dtype=np.int)
    for veci, ij in enumerate(itertools.combinations(range(len(seqs)), 2)):
        indices[veci, :] = ij

    distance_vec(dvec, indices, nb_seqs, nb_metric, *args)
    return dvec

if not NB_SUCCESS:
    nb_hamming_distance = None
    nb_subst_metric = None
else:
    @numba.jit(nopython=True)
    def nb_hamming_distance(str1, str2):
        assert len(str1) == len(str2)

        tot = 0
        for i in range(len(str1)):
            if str1[i] != str2[i]:
                tot += 1
        return tot

    @numba.jit(nopython=True)
    def nb_subst_metric(seq1, seq2, subst_dict, as_similarity=False):
        """Computes sequence similarity based on the substitution matrix."""
        assert len(seq1) == len(seq2)

        def _sim_func(s1, s2, subst):
            sim12 = 0.
            for i in range(len(s1)):
                k1 = s1[i] + '|' + s2[i]
                k2 = s2[i] + '|' + s1[i]
                sim12 += subst.get(k1, subst.get(k2, subst['n|a']))
            return sim12

        """Site-wise similarity between seq1 and seq2 using the substitution matrix subst"""
        sim12 = _sim_func(seq1, seq2, subst_dict)

        if as_similarity:
            return sim12
        else:
            L = len(seq1)
            sim11 = _sim_func(seq1, seq1, subst_dict)
            sim22 = _sim_func(seq2, seq2, subst_dict)
            D = sim11 + sim22 - 2 * sim12
            return D

pwseqdist/matrices.py
import parasail
import numpy as np
import itertools

__all__ = ['parasail_aa_alphabet',
           'seq2vec',
           'vec2seq',
           'seqs2mat',
           'mat2seqs',
           'dict_from_matrix']


"""Parasail uses a 23 AA alphabet for its substitution matrix.
A final column/row includes a value for any symbol not included
in the alphabet"""
parasail_aa_alphabet = 'ARNDCQEGHILKMFPSTWYVBZX'
"""Used for reconstruction of sequences from number representation"""
parasail_aa_alphabet_with_unknown = 'ARNDCQEGHILKMFPSTWYVBZX*'


def dict_from_matrix(mat, alphabet = parasail_aa_alphabet):
    """
    Converts substitution matrix to a dict for use
    by str_subst_metric. Does not store (aa1, aa2) and
    (aa2, aa1), so user needs to check for both. Similarity
    for any symbol not present in the aplhabet is stored in
    key "na" 

    Parameters
    ----------
    mat : parasail substitution Matrix object
        Example is parasail.
    alphabet : str
        string of characters usely specifying the 23 AA Parasail 
        uses for its substitution matrix.

    Returns
    -------
    d : dict
    """
    d = {(aa1, aa2):mat.matrix[alphabet.index(aa1), alphabet.index(aa2)] for aa1, aa2 in itertools.product(alphabet, alphabet)}
    d.update({'na':mat.matrix[-1, 0]})
    return d

def seq2vec(seq, alphabet = parasail_aa_alphabet_with_unknown ):
    """
    Convert AA string sequence into numpy vector of integers

    Parameters
    ----------
    seq : string

    alphabet : str
        string of characters usely specifying the 23 AA Parasail 
        uses for its substitution matrix.

    Returns
    -------
    vec : np.ndarray

    Examples 
    -------_
    >>> seq2vec("CAT")
    array([ 4,  0, 16], dtype=int8)
    >>> seq2vec("ARNDCQEGHILKMFPSTWYVBZX")
    array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22], dtype=int8)
    """
    vec = np.zeros(len(seq), dtype=np.int8)
    for aai, aa in enumerate(seq):
        try:
            vec[aai] =alphabet.index(aa)
        except ValueError("Unknown symbols given value for last column/row of matrix"):
            """Unknown symbols given value for last column/row of matrix"""
            vec[aai] = len(alphabet)
    return vec

def vec2seq(vec, alphabet = parasail_aa_alphabet_with_unknown  ):
    """
    Convert a numpy array of integers back into a AA string sequence.
    (opposite of seq2vec())
    
    Parameters
    ----------
    vec : np.array

    alphabet : str
        string of characters usely specifying the 23 AA Parasail 
        uses for its substitution matrix.

    Returns
    -------
    seq : string

    Examples 
    --------
    >>> vec2seq([ 4,  0, 16])
    'CAT'
    """
    try:
        seq =  ''.join([alphabet[aai] for aai in vec])
    except IndexError:
        max_int = len(alphabet)-1
        raise ValueError(f"vex2seq only works with integers 0 to {max_int} corresponding with {alphabet}")
    return seq

def seqs2mat(seqs):
    """
    Convert a collection of AA sequences into a
    numpy matrix of integers for fast comparison.

    Parameters
    ----------
    seqs : list 
        list of strings of equal length

    Returns
    -------
    mat : np.array  

    Examples
    --------
    >>> seqs2mat(["CAT","HAT"])
    array([[ 4,  0, 16],
           [ 8,  0, 16]], dtype=int8)

    Notes
    -----

    Requires all seqs to have the same length.

    To avoid confusion, mat in this context is an np.array 
    wherase mat in dict_from_matrix() is a parasail substitution Matrix object
    """
    L1 = len(seqs[0])
    mat = np.zeros((len(seqs), L1), dtype=np.int8)
    for si, s in enumerate(seqs):
        msg = "All sequences must have the same length: L1 = %d, but L%d = %d" % (L1, si, len(s))
        assert L1 == len(s), msg
        mat[si, :] = seq2vec(s)
    return mat

def mat2seqs(mat):
    """
    Convert a matrix of integers into AA sequences.
    
    Parameters
    ----------
    mat : np.array 
        2D np.array of integers

    Returns 
    -------
    seqs : list
        list of strings
    
    Examples 
    --------
    >>> mat2seqs(np.array([[ 4,  0, 16],[ 8,  0, 16]] ))
    ['CAT', 'HAT']
    """
    seqs = [vec2seq(mat[i,:]) for i in range(mat.shape[0])]
    return seqs
